src/lib.rs

```diff
@@ -275,14 +282,88 @@ fn parse_unary_operator(input: ParseInput) -> ParseResult<UnaryOperator> {
     spanned(op).parse(input)
 }
 
-/// Parses unary and atomic expressions. E.g., `-u1:1`, `u1:1`
+/// Parses a binary operator. E.g. `|`, `&`, etc.
+fn parse_binary_operator(input: ParseInput) -> ParseResult<BinaryOperator> {
+    let op = alt((
+        // These two must match before | and &
+        value(RawBinaryOperator::BooleanOr, tag("||")),
+        value(RawBinaryOperator::BooleanAnd, tag("&&")),
+        // now match | and &
+        value(RawBinaryOperator::BitwiseOr, tag("|")),
+        value(RawBinaryOperator::BitwiseAnd, tag("&")),
+        value(RawBinaryOperator::BitwiseXor, tag("^")),
+        // concatenate must match before +
+        value(RawBinaryOperator::Concatenate, tag("++")),
+        // the order of these does not matter
+        // arithmetic
+        value(RawBinaryOperator::Add, tag("+")),
+        value(RawBinaryOperator::Subtract, tag("-")),
+        value(RawBinaryOperator::Multiply, tag("*")),
+        // shift
+        value(RawBinaryOperator::ShiftRight, tag(">>")),
+        value(RawBinaryOperator::ShiftLeft, tag("<<")),
+        // compare
+        value(RawBinaryOperator::Equal, tag("==")),
+        value(RawBinaryOperator::NotEqual, tag("!=")),
+        value(RawBinaryOperator::GreaterOrEqual, tag(">=")),
+        value(RawBinaryOperator::Greater, tag(">")),
+        value(RawBinaryOperator::LessOrEqual, tag("<=")),
+        value(RawBinaryOperator::Less, tag("<")),
+    ));
+    spanned(op).parse(input)
+}
+
+/// Parses a let expression. E.g. `let x : u32 = a + u32:1; x`
+///
+/// Note the trailing expression `x`. It is part of the let expression. It is common to have an
+/// expression after the let (after all, what's the point of declaring a variable binding if
+/// you're never going to use the variable?), but not required. So the trailing expression is
+/// optional.
+fn parse_let_expression(
+    input: ParseInput,
+) -> ParseResult<(NonEmpty<LetBinding>, Option<Expression>)> {
+    fn parse_let_binding(input: ParseInput) -> ParseResult<RawLetBinding> {
+        let var_decl = delimited(
+            // let must be followed by at least 1 whitespace
+            tuple((tag_ws("let"), whitespace_exactly1)),
```

Comment:
Does this fail on `let         foo = bar;`?

Reply:
(I didn't see, from a quick scan, that `parse_variable_declaration` -> `parse_identifier` consumed preceding whitespace.)

Reply:
`let         foo = bar;` does not work, I believe because the type annotation is required.

`let  foo : u32 = bar;` (two spaces in there) works fine. See the new test added on line 1558.

```
fn parse_variable_declaration(input: ParseInput) -> ParseResult<VariableDecl> {
    spanned(tuple((
```
`spanned` consumes preceding whitespace. That's how the second space is consumed.